src/analysis/raster/qgskde.cpp

```diff
@@ -90,16 +83,35 @@ QgsKernelDensityEstimation::Result QgsKernelDensityEstimation::prepare()
   int rows = qMax( ceil( mBounds.height() / mPixelSize ) + 1, 1.0 );
   int cols = qMax( ceil( mBounds.width() / mPixelSize ) + 1, 1.0 );
 
-  if ( !createEmptyLayer( driver, mBounds, rows, cols ) )
+  // create empty raster and fill it with nodata values
+  QgsRasterFileWriter writer( mOutputFile );
+  writer.setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer.setOutputFormat( mOutputFormat );
+  mProvider = writer.createOneBandRaster( Qgis::Float32, cols, rows, mBounds, mSource->sourceCrs() );
+  if ( !mProvider )
+  {
     return FileCreationError;
+  }
 
-  // open the raster in GA_Update mode
-  mDatasetH = GDALOpen( mOutputFile.toUtf8().constData(), GA_Update );
-  if ( !mDatasetH )
-    return FileCreationError;
-  mRasterBandH = GDALGetRasterBand( mDatasetH, 1 );
-  if ( !mRasterBandH )
+  if ( !mProvider->setNoDataValue( 1, NO_DATA ) )
+  {
     return FileCreationError;
+  }
+
+  int dataTypeSize = QgsRasterBlock::typeSize( Qgis::Float32 );
+  float *line = new float[ cols ];
+  for ( int i = 0; i < cols; i++ )
```

Comment:
I guess `memset` eould do the same but more efficient?

Reply:
Using std::fill as we already use std::vector. Or better to use `memset` anyway?

Reply:
Ah, right, sounds `std::fill` is the way to go (https://stackoverflow.com/questions/1373369/which-is-faster-preferred-memset-or-for-loop-to-zero-out-an-array-of-doubles#comment1217663_1373422).